src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/TaskResourceSpecification.java

```diff
@@ -0,0 +1,200 @@
+package uk.gov.hmcts.reform.wataskmanagementapi.cft.query;
+
+import org.springframework.data.jpa.domain.Specification;
+import uk.gov.hmcts.reform.wataskmanagementapi.auth.access.entities.AccessControlResponse;
+import uk.gov.hmcts.reform.wataskmanagementapi.auth.permission.entities.PermissionTypes;
+import uk.gov.hmcts.reform.wataskmanagementapi.auth.role.entities.RoleAssignment;
+import uk.gov.hmcts.reform.wataskmanagementapi.cft.entities.TaskResource;
+import uk.gov.hmcts.reform.wataskmanagementapi.cft.entities.TaskRoleResource;
+import uk.gov.hmcts.reform.wataskmanagementapi.cft.enums.CFTTaskState;
+import uk.gov.hmcts.reform.wataskmanagementapi.controllers.request.SearchTaskRequest;
+import uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.search.SearchParameter;
+import uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.search.SearchParameterKey;
+
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+import javax.persistence.criteria.Join;
+import javax.persistence.criteria.Predicate;
+
+@SuppressWarnings({"PMD.DataflowAnomalyAnalysis", "PMD.TooManyMethods", "PMD.LawOfDemeter"})
+public final class TaskResourceSpecification {
+
+    public static final String TASK_ROLE_RESOURCES = "taskRoleResources";
+    public static final String STATE = "state";
+    public static final String LOCATION = "location";
+    public static final ZoneId ZONE_ID = ZoneId.of("Europe/London");
+    public static final String TASK_ID = "taskId";
+    public static final String TASK_TYPE = "taskType";
+    public static final String ASSIGNEE = "assignee";
+    public static final String CASE_ID = "caseId";
+
+    private TaskResourceSpecification() {
+        // avoid creating object
+    }
+
+    public static Specification<TaskResource> getTasks(
+        SearchTaskRequest searchTaskRequest,
+        AccessControlResponse accessControlResponse,
+        List<PermissionTypes> permissionsRequired
+    ) {
+
+        return searchByRoles(permissionsRequired, accessControlResponse)
+            .and(searchByState(searchTaskRequest)
+                .and(searchByLocation(searchTaskRequest)
+                    .and(searchByJurisdiction(searchTaskRequest)
+                        .and(searchByCaseId(searchTaskRequest)
+                            .and(searchByUser(searchTaskRequest)
+                                .and(searchByTaskType(searchTaskRequest)
+                                    .and(searchByTaskId(searchTaskRequest))))))));
```

Comment:
I think the main search endpoint does not allow searching by `taskId` or `taskType`.  So I assume the intention then is to use this method for the search-for-completable endpoint and GET task/{task-id} endpoint?  Then are we going to put the search parameters into the `SearchTaskRequest`?  That may become quite confusing as we are mixing the params from different endpoints.

Reply:
It does allow . The SearchParameterKey has following enum values which user can search on

    LOCATION("location"),
    USER("user"),
    JURISDICTION("jurisdiction"),
    STATE("state"),
    TASK_ID("taskId"),
    TASK_TYPE("taskType"),
    CASE_ID("caseId");